connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala

```diff
@@ -92,17 +92,21 @@ sealed abstract class UserDefinedFunction {
 /**
  * Holder class for a scalar user-defined function and it's input/output encoder(s).
  */
-case class ScalarUserDefinedFunction(
+case class ScalarUserDefinedFunction private (
     function: AnyRef,
     inputEncoders: Seq[AgnosticEncoder[_]],
     outputEncoder: AgnosticEncoder[_],
     name: Option[String],
     override val nullable: Boolean,
-    override val deterministic: Boolean)
+    override val deterministic: Boolean,
```

Comment:
We could also make the serialized `udf` part of the parameters. That should make it a bit easier to work with.

Reply:
Thanks for the suggestion! It is indeed easier to work with it this way. 
I also just realized the previous solution is flawed since `nullable` is part of the proto.ScalarScalaUDF so we can't override it in the case of `asNonNullable` update.

In the new approach, I add the `serializedUdfPacket`  + the input/output proto types so that we can reconstruct the proto.ScalarScalaUDF message without hitting the ser issue while also supporting the nullable update. PTAL